crates/floresta-chain/src/pruned_utreexo/consensus.rs

```diff
@@ -122,33 +126,83 @@ impl Consensus {
+        let mut abort: bool = false;
         // Skip the coinbase tx
         for (n, transaction) in transactions.iter().enumerate() {
+            //We cannot break during value calculation, so we wait until the next iteration.
+            if abort {
+                break;
+            }
             // We don't need to verify the coinbase inputs, as it spends newly generated coins
             if transaction.is_coinbase() {
-                if n == 0 {
-                    continue;
+                if n != 0 {
+                    // A block must contain only one coinbase, and it should be the fist thing inside it
+                    return Err(BlockValidationErrors::FirstTxIsnNotCoinbase.into());    
+                }
+                //the prevout input of a coinbase must be all zeroes
+                if transaction.input[0].previous_output.txid != Txid::all_zeros() { return Err(BlockValidationErrors::InvalidCoinbase("Invalid coinbase txid".to_string()).into());}
+                let scriptsig =  transaction.input[0].script_sig.clone();
+                let scriptsigsize = scriptsig.clone().into_bytes().len();
+                if  scriptsigsize > 100 || scriptsigsize < 2 {
+                    //the scriptsig size must be between 2 and 100 bytes
+                    return Err(BlockValidationErrors::InvalidCoinbase("Invalid ScriptSig size".to_string()).into());
                 }
-                // A block must contain only one coinbase, and it should be the fist thing inside it
-                return Err(BlockValidationErrors::FirstTxIsnNotCoinbase.into());
+                //according BIP 34, the scriptsig of a coinbase must be the height of the block;
+                //the first byte is the number of bytes of the height
+                let height_index = u8::from_le(scriptsig.as_script().as_bytes()[0]);
+                //gets the height in bytes (including sign)
+                let in_script_height = &scriptsig.as_script().as_bytes()[1..height_index as usize];
+                //height from le to i64
+                let in_script_height = i32::from_le_bytes(in_script_height.try_into().unwrap()); 
+                if in_script_height != height{
+                    return Err(BlockValidationErrors::InvalidCoinbase("Invalid declared Block Height in ScriptSig".to_string()).into());
+                };
+                continue;
+                
+                
             }
             // Amount of all outputs
-            let output_value = transaction
-                .output
-                .iter()
-                .fold(0, |acc, tx| acc + tx.value.to_sat());
+            let mut output_value = 0;
+            for output in transaction.output.iter() {
+                if output.value.to_sat() > 0 {
+                    output_value += output.value.to_sat()
+                } else {
+                    return Err(BlockValidationErrors::InvalidTx(alloc::format!(
+                        "Invalid output: {:?}",
```

Comment:
The same for this kind of operation and other ones that can be delegated to a function that returns ok(()) if is validated.

Reply:
> The same for this kind of operation and other ones that can be delegated to a function that returns ok(()) if is validated.



> The same for this kind of operation and other ones that can be delegated to a function that returns ok(()) if is validated.

I would prefer that, I usually hate nesting.